roles/mining-proxy/src/lib/downstream_mining.rs

```diff
@@ -487,40 +487,49 @@ use network_helpers_sv2::plain_connection_tokio::PlainConnection;
 use std::net::SocketAddr;
 use tokio::net::TcpListener;
 
-pub async fn listen_for_downstream_mining(address: SocketAddr) {
+pub async fn listen_for_downstream_mining(address: SocketAddr, mut shutdown_rx: TokioReceiver<()>) {
     info!("Listening for downstream mining connections on {}", address);
-    let listner = TcpListener::bind(address).await.unwrap();
+    let listener = TcpListener::bind(address).await.unwrap();
     let mut ids = roles_logic_sv2::utils::Id::new();
 
-    while let Ok((stream, _)) = listner.accept().await {
-        let (receiver, sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
-            PlainConnection::new(stream).await;
-        let node = DownstreamMiningNode::new(receiver, sender, ids.next());
-
-        task::spawn(async move {
-            let mut incoming: StdFrame = node.receiver.recv().await.unwrap().try_into().unwrap();
-            let message_type = incoming.get_header().unwrap().msg_type();
-            let payload = incoming.payload();
-            let routing_logic = super::get_common_routing_logic();
-            let node = Arc::new(Mutex::new(node));
-
-            // Call handle_setup_connection or fail
-            match DownstreamMiningNode::handle_message_common(
-                node.clone(),
-                message_type,
-                payload,
-                routing_logic,
-            ) {
-                Ok(SendToCommon::RelayNewMessageToRemote(_, message)) => {
-                    let message = match message {
-                        roles_logic_sv2::parsers::CommonMessages::SetupConnectionSuccess(m) => m,
+    let mut should_continue = true;
+    while should_continue {
```

Comment:
Could we use `loop` and `break` in the `Err` arm?

Reply:
I tried that first, but as the code that listens for the shutdown message is inside a macro, it can't break from there. At least I couldn't get it to compile with `break` from inside `select!`, not even using named loops